### ISSUE TITLE
CTGAN generator model - input also connect to layer2 ?

### ISSUE THREAD

**Issue by petercoppensdatylon** [sdv-dev/SDV#839]
Hello,

Disclaimer : nn and gan beginner

Now...I am looking at the generator network of  a CTGAN where the Input is 3 continuous variables, configured with 2 layers of 256 nodes each. Model is here

![image](https://user-images.githubusercontent.com/79080548/172896632-9452744d-eded-40e2-857a-44b405855e37.png)

I sort of get it from the dump, but one thing that strikes me as unexpected is the fact that the inputs also seem to be connected to the layer 2 nodes. Not sure, but is that no atypical for neural networks in general?

I also cannot seem to distill it from the model in "Modeling Tabular Data using Conditional GAN" paper (although I do not fully understand the notation there)

![image](https://user-images.githubusercontent.com/79080548/172897338-8b131c96-7d03-4ba8-80d5-86d541c3e282.png)

All tips or pointer to info on this are welcomed warmly!

Wkr,

Peter

**Comment by fealho** [comment]
Hi @petercoppensdatylon, you mention that "the inputs also seem to be connected to the layer 2 nodes.". Could you explain how you concluded that from the dump?

**Comment by petercoppensdatylon** [comment]
Hello

Thanks for your reaction.

I was trying to understand the node counts for the total network (and intermediate layers) in the dump and could only make it match using that assumption

Like so
- Layer 1: 256 * 128 + 256 --> 33024
- Normalization: 512 --> 512
- Layer 2:  256 * 256 + 256 + 128 * 256 --> 98560
- Normalization: 512 --> 512
- Output: 256 * 28 + 28 + 28 * ( 128 + 256 ) --> 17948

Where the '128 * 256' and  '28 * ( 128 + 256 )' parts would then be the connects from 'earlier' layers.

Like this, it adds up to the 150556 from the dump

It is sort of confirmed by looking at the Generator object in the debugger as well through the 'in_features=384'  part of the second layer and the 'in_features=640' of the last layer

self._generator
-->
```
Generator(
  (seq): Sequential(
    (0): Residual(
      (fc): Linear(in_features=128, out_features=256, bias=True)
      (bn): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU()
    )
    (1): Residual(
      (fc): Linear(in_features=384, out_features=256, bias=True)
      (bn): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU()
    )
    (2): Linear(in_features=640, out_features=28, bias=True)
  )
)

```

Finally the code has

```
    def __init__(self, embedding_dim, generator_dim, data_dim):
        super(Generator, self).__init__()
        dim = embedding_dim
        seq = []
        for item in list(generator_dim):
            seq += [Residual(dim, item)]
            dim += item
        seq.append(Linear(dim, data_dim))
        self.seq = Sequential(*seq)

```
Where the seq += part in the loop sort of explains that as well (although, again with the disclaimer I am not really grasping what is going on here in detail - I am not a Python programmer nor a pytorch user)

I assume my reasoning is flawed somewhere but I am not sure how else to explain the node counts for the Generator network.

Wkr,

Peter

**Comment by fealho** [comment]
@petercoppensdatylon thanks for the clarification, your understanding is indeed correct. The input is concatenated to the feature representation - this is a variant of resnet which uses concatenation to mitigate the vanishing gradients problem (instead of the addition in the original paper). 

Here are some references:

[1] https://arxiv.org/pdf/1512.03385.pdf
[2] https://towardsdatascience.com/an-overview-of-resnet-and-its-variants-5281e2f56035

**Comment by npatki** [comment]
Hi @petercoppensdatylon, seems like @fealho answered your Q so I'll close off the issue.

Feel free to respond and re-open if there is anything else related to this topic.
